### PR TITLE
continuation of image_ref work

### PR DESCRIPTION
**Pull request by bparees**
- [x] make From a pointer

### CODE REVIEW

**Comment by bparees** [comment]
@mfojtik fyi.

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
@mfojtik @smarterclayton there is a fairly ugly consequence to this change...  specifically, buildConfigs are defined to prefer their "From" field but use the "Image" field if From is not present.  (in fact it's a validation error to supply both, currently).  However, Build objects (which use the same STIStrategy type as a field) logically require that you *only* supply an Image.  And even if we allowed you to supply a From field to a Build (on the assumption we'd resolve the From reference as we created the build pod), we'd still have to prefer the Image field since the Image field might have been set earlier by the Tag resolution logic.  So it would have the inverse precedence of the BuildConfig.  

For now i've documented it as saying that Build objects only use the Image field.  I don't really want to create a second STIStrategy type (that only has an Image field), but as things stand now, Build objects are a little funky.

My assumption/justificatin/hope for now is that for the most part people will be creating BuildConfigs, not Builds (especially once we have @soltysh 's new build creation endpoints) and thus people won't be exposed to the Build object semantics.

And i do think it's better that Builds only have Image, not From references, because it makes them more re-runnable.  Adding something that's dynamically resolved (like an ImageRepo reference) makes them significantly less immutable.

**Comment by bparees** [comment]
@mfojtik @soltysh ptal.  the patterns here are going to be reproduced for Docker and Custom build strategies.

File: pkg/build/api/v1beta1/conversion.go [hunk 1]
```diff
@@ -53,11 +53,25 @@ func init() {
 		func(in *newer.STIBuildStrategy, out *STIBuildStrategy, s conversion.Scope) error {
 			out.BuilderImage = in.Image
 			out.Image = in.Image
+			if in.From != nil {
+				out.From = &kapi.ObjectReference{
+					Name:      in.From.Name,
```
Comment (mfojtik) [inline pkg/build/api/v1beta1/conversion.go]: If this is a ImageRepository reference, can you set the Kind as well?

File: pkg/build/util/generate.go [hunk 2]
```diff
@@ -13,7 +14,8 @@ import (
 // GenerateBuildFromConfig creates a new build based on a given BuildConfig. Optionally a SourceRevision for the new
 // build can be specified.  Also optionally a list of image names to be substituted can be supplied.  Values in the BuildConfig
 // that have a substitution provided will be replaced in the resulting Build
-func GenerateBuildFromConfig(bc *buildapi.BuildConfig, r *buildapi.SourceRevision, imageSubstitutions map[string]string) (build *buildapi.Build) {
+func GenerateBuildFromConfig(bc *buildapi.BuildConfig, r *buildapi.SourceRevision, imageSubstitutions map[string]string,
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: s/r/ref/

File: pkg/build/util/generate.go [hunk 3]
```diff
@@ -155,3 +219,18 @@ func SubstituteImageReferences(build *buildapi.Build, oldImage string, newImage
 		}
 	}
 }
+
+// SubstituteImageRepoReferences uses references to an image repository to set an actual image name
+// It also clears the ImageRepo reference from the BuildStrategy, if one was set.  The imagereference
+// field will be used explicitly.
+func SubstituteImageRepoReferences(build *buildapi.Build, imageRepo kapi.ObjectReference, newImage string) {
+	switch {
+	case build.Parameters.Strategy.Type == buildapi.STIBuildStrategyType &&
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: this is pretty ugly case :)

File: pkg/build/util/generate.go (hunk 2 above)
Comment (bparees) [inline pkg/build/util/generate.go]: you know that isn't even something this pull introduced, right? :)

File: pkg/build/util/generate.go [hunk 4]
```diff
@@ -62,10 +72,63 @@ func GenerateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
...
+			build, err = GenerateBuildUsingObjectReference(config, revision, imageRepoGetter)
+		} else {
+			build, err = GenerateBuildUsingImageTriggerTag(config, revision, imageRepoGetter)
+		}
+	case config.Parameters.Strategy.Type == buildapi.DockerBuildStrategyType:
+		build, err = GenerateBuildUsingImageTriggerTag(config, revision, imageRepoGetter)
+	case config.Parameters.Strategy.Type == buildapi.CustomBuildStrategyType:
+		build, err = GenerateBuildUsingImageTriggerTag(config, revision, imageRepoGetter)
+	default:
+		err = fmt.Errorf("Build strategy type must be set")
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: why not just return error here?

File: pkg/build/api/v1beta1/conversion.go (hunk 1 above)
Comment (bparees) [inline pkg/build/api/v1beta1/conversion.go]: set

File: pkg/build/util/generate.go [hunk 5]
```diff
@@ -62,10 +72,63 @@ func GenerateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
...
+	}
+	return build, err
+}
+
+func GenerateBuildUsingObjectReference(config *buildapi.BuildConfig, revision *buildapi.SourceRevision, imageRepoGetter osclient.ImageRepositoryNamespaceGetter) (*buildapi.Build, error) {
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: godoc

File: pkg/build/util/generate.go (hunk 2 above)
Comment (bparees) [inline pkg/build/util/generate.go]: renamed

File: pkg/build/util/generate.go [hunk 6]
```diff
@@ -62,10 +72,63 @@ func GenerateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
...
+	namespace := from.Namespace
+	if len(namespace) == 0 {
+		namespace = config.Namespace
+	}
+	tag := config.Parameters.Strategy.STIStrategy.Tag
+	if len(tag) == 0 {
+		tag = buildapi.DefaultImageTag
+	}
+
+	var imageRepo *imageapi.ImageRepository
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: why declaring these by var? you can just do imageRepo, err :=

File: pkg/build/util/generate.go [hunk 7]
```diff
@@ -62,10 +72,63 @@ func GenerateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
...
+	glog.V(4).Infof("Found image repo %s", imageRepo.Name)
+	latest, err := imageapi.LatestTaggedImage(imageRepo, tag)
+	if err == nil {
+		glog.V(4).Infof("Using image %s for image repository %s in namespace %s", latest.DockerImageReference, from.Name, from.Namespace)
+		imageRepoSubstitutions[*from] = latest.DockerImageReference
+	} else {
+		return nil, fmt.Errorf("Docker Image Repository %s has no tag %s", from.Name, tag)
+	}
+	glog.V(4).Infof("Generating build from config for build config %s", config.Name)
+	build := GenerateBuildFromConfig(config, revision, nil, imageRepoSubstitutions)
```
Comment (mfojtik) [inline pkg/build/util/generate.go]: nuke 'build' var and just return GenerateBuildConfig(), nil

File: pkg/build/util/generate.go (hunk 3 above)
Comment (bparees) [inline pkg/build/util/generate.go]: it's beautiful on the inside.

**Comment by bparees** [comment]
@mfojtik comments addressed, thanks

**Comment by bparees** [comment]
fixes https://github.com/openshift/origin/issues/1418

File: pkg/build/util/generate.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/util/generate.go]: I'm not in favor of this and couple others like this, but I can't think of anything better, yet :wink:

**Comment by soltysh** [comment]
LGTM once it's rebased and passes the tests. Good job @bparees :8ball:

**Comment by bparees** [comment]
@soltysh mostly not my work but thanks :/

**Comment by soltysh** [comment]
Just got confirmation from @mfojtik he's OK with it, thus [merge]-ing.

**Comment by smarterclayton** [comment]
I've rolled this back because it was randomly failing integration tests and end to end.

**Comment by smarterclayton** [comment]
Please resubmit

**Comment by smarterclayton** [comment]
```
ok      TestSelfSubjectAccessReviews
failed  TestSimpleConfigChangeTrigger
W0324 13:03:18.122659   18925 master.go:186] Portal net unspecified. Defaulting to 10.0.0.0/24.
I0324 13:03:18.122739   18925 util.go:410] Choosing interface en0 for from-host portals
I0324 13:03:18.122754   18925 util.go:415] Interface en0 = fe80::7a31:c1ff:feb7:6042/64
I0324 13:03:18.122762   18925 master.go:214] Will report fe80::7a31:c1ff:feb7:6042 as public IP address.
I0324 13:03:18.122777   18925 master.go:260] Setting master service IPs based on PortalNet subnet to "10.0.0.1" (read-only) and "10.0.0.2" (read-write).
I0324 13:03:18.160407   18925 controller.go:40] Waiting for first version of integration/change-deploy-config:0
I0324 13:03:18.208057   18925 controller.go:70] Deployment already exists for config integration/change-deploy-config:1
I0324 13:03:18.210811   18925 controller.go:55] Created initial deployment for config integration/change-deploy-config:0
I0324 13:03:18.250687   18925 controller.go:75] Ignoring config change for integration/change-deploy-config:1 (latestVersion=1); same as deployment integration/change-deploy-config-1
I0324 13:03:18.319873   18925 controller.go:86] Updated config integration/change-deploy-config:1 from version 1 to 2 for existing deployment integration/change-deploy-config-1
I0324 13:03:18.321995   18925 controller.go:63] Ignoring config change for integration/change-deploy-config:2; no existing deployment found
--- FAIL: TestSimpleConfigChangeTrigger (0.23s)
	deploy_trigger_test.go:325: Starting test openshift
	deploy_trigger_test.go:285: expected watch event type ADDED, got MODIFIED
FAIL
failed  TestSimpleImageChangeBuildTrigger
W0324 13:03:18.554320   18932 master.go:186] Portal net unspecified. Defaulting to 10.0.0.0/24.
I0324 13:03:18.554439   18932 util.go:410] Choosing interface en0 for from-host portals
I0324 13:03:18.554462   18932 util.go:415] Interface en0 = fe80::7a31:c1ff:feb7:6042/64
I0324 13:03:18.554474   18932 master.go:214] Will report fe80::7a31:c1ff:feb7:6042 as public IP address.
I0324 13:03:18.554491   18932 master.go:260] Setting master service IPs based on PortalNet subnet to "10.0.0.1" (read-only) and "10.0.0.2" (read-write).
I0324 13:03:18.604712   18932 image_change_controller.go:39] Build image change controller detected imagerepo change registry:8080/openshift/test-image-trigger
I0324 13:03:18.604726   18932 image_change_controller.go:83] Running build for buildConfig test-build-cfg in namespace integration
I0324 13:03:18.605470   18932 generate.go:39] Substituting registry:8080/openshift/test-image-trigger:latest for registry:8080/openshift/test-image-trigger
I0324 13:03:18.606063   18932 controller.go:101] Updated all configs for trigger on imageRepo integration/test-image-trigger-repo
I0324 13:03:18.610350   18932 controller.go:42] Handling build test-build-cfg-ab46c96d-d247-11e4-a9ec-7831c1b76042
--- FAIL: TestSimpleImageChangeBuildTrigger (0.12s)
	deploy_trigger_test.go:325: Starting test openshift
	imagechange_buildtrigger_test.go:72: expected watch event type MODIFIED, got ADDED
FAIL
failed  TestSimpleImageChangeTrigger
W0324 13:03:18.709405   18939 master.go:186] Portal net unspecified. Defaulting to 10.0.0.0/24.
I0324 13:03:18.709488   18939 util.go:410] Choosing interface en0 for from-host portals
I0324 13:03:18.709502   18939 util.go:415] Interface en0 = fe80::7a31:c1ff:feb7:6042/64
I0324 13:03:18.709510   18939 master.go:214] Will report fe80::7a31:c1ff:feb7:6042 as public IP address.
I0324 13:03:18.709525   18939 master.go:260] Setting master service IPs based on PortalNet subnet to "10.0.0.1" (read-only) and "10.0.0.2" (read-write).
I0324 13:03:18.743417   18939 image_change_controller.go:39] Build image change controller detected imagerepo change registry:8080/openshift/test-image
I0324 13:03:18.745286   18939 controller.go:101] Updated all configs for trigger on imageRepo integration/test-image-repo
I0324 13:03:18.756752   18939 controller.go:40] Waiting for first version of integration/image-deploy-config:0
I0324 13:03:18.757980   18939 controller.go:43] Ignoring config integration/image-deploy-config:0; no change triggers detected
I0324 13:03:18.778111   18939 controller.go:43] Ignoring config integration/image-deploy-config:1; no change triggers detected
I0324 13:03:18.803168   18939 image_change_controller.go:39] Build image change controller detected imagerepo change registry:8080/openshift/test-image
I0324 13:03:18.805002   18939 controller.go:38] Detecting changed images for deploymentConfig integration/image-deploy-config:1
I0324 13:03:18.805025   18939 controller.go:48] Found matching ImageChange trigger for deploymentConfig integration/image-deploy-config:1: &api.DeploymentTriggerImageChangeParams{Automatic:true, ContainerNames:[]string{"container-1"}, RepositoryName:"registry:8080/openshift/test-image", From:api.ObjectReference{Kind:"", Namespace:"", Name:"", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}, Tag:"latest"}
I0324 13:03:18.805076   18939 controller.go:54] Processing image triggers for deploymentConfig integration/image-deploy-config:1
I0324 13:03:18.805098   18939 controller.go:79] Container container-1 for config integration/image-deploy-config:1: image id changed from "ref-1" to ""; regenerating config
I0324 13:03:18.866738   18939 controller.go:43] Ignoring config integration/image-deploy-config:2; no change triggers detected
I0324 13:03:18.870061   18939 controller.go:94] Updated deploymentConfig integration/image-deploy-config:1 in response to image change trigger
I0324 13:03:18.870071   18939 controller.go:101] Updated all configs for trigger on imageRepo integration/test-image-repo
--- FAIL: TestSimpleImageChangeTrigger (0.19s)
	deploy_trigger_test.go:325: Starting test openshift
	deploy_trigger_test.go:159: expected watch event type ADDED, got MODIFIED
FAIL
ok      TestSimpleImageChangeTriggerFrom
```
was seen local, and the last two merge e2e's failed (after this landed)

**Comment by bparees** [comment]
only one merge e2e failed since this landed and it was due to the DB pod not starting.

**Comment by smarterclayton** [comment]
When is this going to be added to Docker and Custom build strategies?  What card / pull tracks that work?

**Comment by bparees** [comment]
There's a card on the next list, it's high on my priorities but I was waiting for the imagestream changes to land so I didn't have to deal with a bunch of rebases. Hoping to start it this week but I also have the PHP image on my list and a variety of other tidbits. 

Ben Parees | OpenShift

**Comment by smarterclayton** [comment]
K.  with the status stuff I'm deep in the guts of using the objects programmatically and I'm seeing the faults.

**Comment by bparees** [comment]
Seeing which faults? 

Ben Parees | OpenShift

**Comment by smarterclayton** [comment]
In the approachability and usability of the APIs.  Same stuff we were discussing last week, difficult to write clean logic for both triggers and builds (and then working with deployments and having opposite behavior for triggers now).  The objects are complex, but we don't have common iteration logic.  

I'm thinking our trigger object references should be to ImageStreamTags now - which would remove the need for the separate tag field.  We might also want to fold docker image references directly into a virtual Kind and treat them as object references.  All things we should discuss as part of our plan for v1.

**Pull request merged**